polynote-server/src/main/scala/polynote/server/NotebookManager.scala

```diff
@@ -0,0 +1,65 @@
+package polynote.server
+
+import java.io.File
+import java.util.concurrent.ConcurrentHashMap
+
+import cats.Monad
+import cats.effect.{ContextShift, Fiber, IO}
+import cats.effect.concurrent.Semaphore
+import polynote.server.repository.NotebookRepository
+
+/**
+  * This is how the sessions interact with notebooks. The sessions don't think about writing changes etc. They'll
+  * open a [[SharedNotebook]] and from there get a [[NotebookRef]] in exchange for an output queue. Their interactions
+  * with the notebook will go through the NotebookRef.
+  */
+abstract class NotebookManager[F[_]](implicit F: Monad[F]) {
+
+  def getNotebook(path: String): F[SharedNotebook[F]]
+
+  def listNotebooks(): F[List[String]]
+
+  def createNotebook(path: String): F[String]
+
+}
+
+class IONotebookManager(
+  repository: NotebookRepository[IO],
+  kernelFactory: KernelFactory[IO])(implicit
+  contextShift: ContextShift[IO]
+) extends NotebookManager[IO] {
+
+  private val writers = new ConcurrentHashMap[String, Fiber[IO, Unit]]
+  private val notebooks = new ConcurrentHashMap[String, SharedNotebook[IO]]
+  private val loadingNotebook = Semaphore[IO](1).unsafeRunSync()
+
+  private def writeChanges(notebook: SharedNotebook[IO]): IO[Fiber[IO, Unit]] = notebook.versions.map(_._2).evalMap {
+    updated => repository.saveNotebook(notebook.path, updated)
```

Comment:
I needed to add the `handleErrorWith` in the old PR (https://github.com/polynote/polynote/pull/3/files) in order to see errors that we ran into while writing the file. Are those errors now handled somewhere else?

Reply:
Good point, I must have lost that change. I'll put it back.